### PR TITLE
Make Recovery API support `detailed` params

### LINKED ISSUE #28910: Recovery API supports `details` param not `detailed`
As part of #28878 we discovered that the recovery API accepts a `detailed` parameter, also documented, that is unused. The parameter should rather be called `details` (see https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java#L918), although it's no longer accepted by the recovery API since we introduced parameters validation. Also the `detailed` flag is transported to the response, without being serialized nor printed out, where it should be removed.

We should make this work, by ~~either accepting the `details` param instead and updating docs and spec or~~ modifying `RecoveryState` to look for `detailed` instead of `details`. The detailed flag though doesn't need to be part of the response, and does not need to be part of the request either, given that `RestRequest` is provided as `Params` argument to the `toXContent` method.

### PR DESCRIPTION
**Pull request by PnPie**
Properly forward the `detailed` parameter to show extra `details` in the recovery stats.

Close #28910

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
  */
 public class RecoveryRequest extends BroadcastRequest<RecoveryRequest> {
 
-    private boolean detailed = false;       // Provides extra details in the response
+    private boolean details = false;       // Provides extra details in the response
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java]: can we remove this too?

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java [hunk 2]
```diff
@@ -91,14 +82,14 @@ public void activeOnly(boolean activeOnly) {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         super.writeTo(out);
-        out.writeBoolean(detailed);
+        out.writeBoolean(details);
```
Comment (bleskes) [inline server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java]: when you remove this, watch out for BWC aspects - see https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/action/admin/cluster/storedscripts/GetStoredScriptRequest.java#L59 as an example.

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx for picking this up. I left some comments

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryResponse.java [hunk 3]
```diff
@@ -39,7 +39,6 @@
  */
 public class RecoveryResponse extends BroadcastResponse implements ToXContentFragment {
 
-    private boolean detailed = false;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryResponse.java]: wasn't this already removed in another PR? Maybe you need to merge master in?

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java (hunk 1 above)
Comment (PnPie) [reply server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java]: The `details` variable in `RecoveryRequest` ? It's to pass the `details` flag in Recovery API ? can we remove it ?

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java (hunk 2 above)
Comment (PnPie) [reply server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequest.java]: Here I just changed the name of variable (to be the same as flag name), the value is always a boolean, so we don't need to do some bwc changes ?

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryResponse.java (hunk 3 above)
Comment (PnPie) [reply server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryResponse.java]: Yes, I didn't merge master before.

**Review by PnPie — COMMENTED** [review]
Hello @bleskes @javanna I update it and added some comments.

**Comment by pcsanwald** [comment]
@javanna @bleskes we discussed this in FixIt Thursday, any thoughts on what is best, here?

**Comment by ywelsch** [comment]
@javanna @bleskes ping

**Comment by javanna** [comment]
hi @PnPie sorry it took ages to get back to you. We have discussed this with the team and we think that we should rather change the code to expose `detailed` like the docs mention, rather than accepting `details`. This is more consistent with other API that already accept `detailed`. Would you be up for updating your PR and merging master in, or opening a new one given all the changes that went in meanwhile? Thanks!

**Comment by PnPie** [comment]
Hi @javanna thanks for your reply and I'm sorry I just get occupied these days and I'll update it during this week-end ;)

**Comment by javanna** [comment]
no problem at all @PnPie we made you wait much longer than that ;)

**Comment by elasticcla** [comment]
Hi @PnPie, we have found your signature in our records, but it seems like you have signed with a different e-mail than the one used in yout Git [commit](https://github.com/elastic/elasticsearch/pull/29076.patch). Can you please add both of these e-mails into your Github profile (they can be hidden), so we can match your e-mails to your Github profile?

**Comment by PnPie** [comment]
Hi @javanna I updated it and added a test case for `RecoveryState`, sorry for the slight delay :)

**Comment by ywelsch** [comment]
@elasticmachine retest this please

File: server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequestBuilder.java [hunk 4]
```diff
@@ -43,4 +43,4 @@ public RecoveryRequestBuilder setActiveOnly(boolean activeOnly) {
         request.activeOnly(activeOnly);
         return this;
     }
-}
\ No newline at end of file
+}
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/action/admin/indices/recovery/RecoveryRequestBuilder.java]: undo this change

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 5]
```diff
@@ -343,7 +343,7 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         static final String REUSED = "reused";
         static final String REUSED_IN_BYTES = "reused_in_bytes";
         static final String PERCENT = "percent";
-        static final String DETAILS = "details";
+        static final String DETAILED = "detailed";
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: let's leave this field as is. "details" in the output is nicer, and we don't break existing parsers.

File: server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java [hunk 6]
```diff
@@ -929,8 +929,8 @@ public synchronized XContentBuilder toXContent(XContentBuilder builder, Params p
             builder.field(Fields.REUSED, reusedFileCount());
             builder.field(Fields.RECOVERED, recoveredFileCount());
             builder.field(Fields.PERCENT, String.format(Locale.ROOT, "%1.1f%%", recoveredFilesPercent()));
-            if (params.paramAsBoolean("details", false)) {
-                builder.startArray(Fields.DETAILS);
+            if (params.paramAsBoolean("detailed", false)) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java]: this is the only change that we need to do in this class

File: server/src/test/java/org/elasticsearch/indices/recovery/RecoveryStateTests.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Collections;
+
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.emptySet;
+
+public class RecoveryStateTests extends ESTestCase {
+
+    public void testRecoveryStateIndexToXContent() {
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/recovery/RecoveryStateTests.java]: I would prefer a REST test to show that the `detailed` flag is honoured (see `rest-api-spec/test/cat.recovery/10_basic.yml`). This test here is not needed.

**Review by ywelsch — CHANGES_REQUESTED** [review]
Thanks for updating this @PnPie. I've left some comments. We should also update the REST spec (`rest-api-spec/api/cat.recovery.json`) to include this flag.

**Comment by PnPie** [comment]
Hi @ywelsch I updated the PR and added a rest layer test.

**Comment by ywelsch** [comment]
@elasticmachine update-branch

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.recovery/10_basic.yml [hunk 8]
```diff
@@ -130,3 +130,25 @@
         index: [v*]
 
   - match: { $body: {} }
+---
+"Indices recovery test with detailed parameter":
```
Comment (ywelsch) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.recovery/10_basic.yml]: I think we will have to skip this test for versions that don't have this fix.

**Review by ywelsch — APPROVED** [review]
@PnPie sorry that I missed the ping here. Thank you for iterating with me on this. I think we will need to make the test conditional on the ES version that it will run against. I've started CI on this, and if necessary, can make the adjustment myself as well.

**Comment by ywelsch** [comment]
@elasticmachine update-branch

**Comment by ywelsch** [comment]
@elasticmachine retest this please

**Comment by ywelsch** [comment]
@elasticmachine retest this please

**Comment by ywelsch** [comment]
@elasticmachine retest this please

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/1
